### PR TITLE
blob: harden async benchmark harness + document GCP connection-pool tuning

### PR DESCRIPTION
**Pull request by iamabhilaksh**
## What & why

While investigating the AWS-vs-GCP async blob throughput gap (W-23487831), I found the async blob **benchmark harness itself** had a correctness bug that would silently produce false results, plus a coverage-agent interference that distorted every measurement. This PR fixes the harness, adds the knobs needed to run controlled provider-tuning experiments, and documents the GCP connection-pool lever that the experiment surfaced.

**No SDK default is changed.** This is a benchmarking + harness-correctness + documentation PR. The two GCP tuning knobs referenced already exist and remain unset by default.

## Changes

1. **Fix: forked-JVM property forwarding (the real bug).** `runBenchmarks` only forwarded `BLOB_BENCHMARK_*` system properties into the forked JMH JVM. `buildStore()` runs *in the fork*, so any `bench.gcp.*` tuning toggle set on the Maven JVM was silently dropped — with `@Fork>=1` a "tuned" leg would run with defaults, making before/after legs identical and yielding a false "no effect." Now forwards `bench.gcp.*` as well.

2. **Parameterize the runner** (all default to prior behavior when unset): `bench.method` (isolate a benchmark), `bench.forks`, `bench.tag` (separate result files), `bench.mode` (single JMH mode instead of paying for every declared mode), and warmup/measurement timing overrides for slow directory ops that need longer iterations for a tight CI.

3. **pom: auto-skip JaCoCo when `runBenchmarks=true`.** The coverage agent instruments the forked JMH JVM and distorts throughput (~40% observed). A `benchmark` profile auto-activates on the property so a benchmark run is never silently measured with the agent attached. No effect on normal `test`/`verify`.

4. **`GcpAsyncBlobBenchmarkTest`: `-Dbench.gcp.tuned` toggle** (default off) to raise the Apache per-route connection cap and the TransferManager worker pool for controlled experiments.

5. **`GcpBlobStore`: comment-only** documentation (no behavior change) explaining that GCS traffic maps to a single Apache HTTP route capped at 20 connections by default while the TransferManager spawns 2×CPU workers, so directory / many-small-object workloads should raise `withMaxConnections` + `withTransferManagerThreadPoolSize` together.

## Validation

Controlled before/after on live GCS, same host/JVM, back-to-back, `@Fork(3)`, Throughput, 99.9% CI, JaCoCo off:

| Benchmark | Before ops/s (99.9% CI) | After ops/s (99.9% CI) | Δ | CIs separated? |
|---|---|---|---|---|
| UploadDirectorySmall | 0.482 [0.451, 0.512] | 1.473 [1.009, 1.937] | **+205.8%** | YES |
| DownloadDirectorySmall | 0.903 [0.885, 0.921] | 2.537 [1.497, 3.576] | **+181.0%** | YES |
| UploadSmall (control, single round trip) | 4.502 [3.319, 5.685] | 4.987 [3.709, 6.265] | +10.8% | no (overlap) |
| DownloadSmall (control, single round trip) | 4.834 [3.257, 6.411] | 5.336 [3.920, 6.752] | +10.4% | no (overlap) |

The two directory ops tripled with cleanly separated CIs; the two single-object controls barely moved — each issues a single HTTP round trip (no TransferManager fan-out), so at 4 benchmark threads they never approach the 20-connection route cap and raising it cannot help them — ruling out a global warmup/JVM artifact and confirming the effect is specific to the parallel directory path. This is evidence the harness now *measures correctly* and the documented lever works; it is **not** a claim of AWS parity (the CRT stack retains an edge) and it does **not** change any default.

## Scope / non-goals

- The GCP **defaults are intentionally not changed**. A global cap/worker bump would enlarge the connection + thread footprint for every GCP caller — penalizing the common single-object / constrained-container case to benefit only many-core directory workloads. The lever stays opt-in via `withMaxConnections` + `withTransferManagerThreadPoolSize`, documented in `GcpBlobStore`. Note the validated gain is host-size dependent (measured on a 16-CPU host; ≤10-CPU containers don't saturate the default cap and see little benefit).
- Complementary to #533 (connection-reaper / TCP keepalive for tail latency) and #532 (saturation metrics) — different levers; no overlap with the per-route cap or worker pool.

Relates to W-23487831.

### CODE REVIEW

File: blob/blob-client/src/test/java/com/salesforce/multicloudj/blob/async/client/AbstractAsyncBlobBenchmarkTest.java [hunk 1]
```diff
@@ -679,20 +679,67 @@ private static void deleteLocalRecursive(Path root) {
...
+        (methodFilter != null && !methodFilter.isEmpty())
+            ? ".*" + this.getClass().getName() + "\\." + methodFilter + ".*"
+            : ".*" + this.getClass().getName() + ".*";
+
+    String resultFile =
+        (resultTag != null && !resultTag.isEmpty())
+            ? "target/jmh-async-results-" + getProviderId() + "-" + resultTag + ".json"
+            : "target/jmh-async-results-" + getProviderId() + ".json";
+
+    OptionsBuilder builder = new OptionsBuilder();
```
Comment (sharatchandrag) [inline blob/blob-client/src/test/java/com/salesforce/multicloudj/blob/async/client/AbstractAsyncBlobBenchmarkTest.java]: Can you improve the code readability in these blocks, If you have regular expressions,  can you also explore using Pattern. quote() that can help to avoid the edge cases

**Review by sharatchandrag — CHANGES_REQUESTED** [review]

File: blob/blob-client/src/test/java/com/salesforce/multicloudj/blob/async/client/AbstractAsyncBlobBenchmarkTest.java (hunk 1 above)
Comment (iamabhilaksh) [reply blob/blob-client/src/test/java/com/salesforce/multicloudj/blob/async/client/AbstractAsyncBlobBenchmarkTest.java]: Good catch — done. Extracted the option-building into named helpers (`collectForwardedProperties`, `buildIncludePattern`, `buildResultFilePath`, `applyTimingOverrides`, `applyModeOverride`) so `runBenchmarks()` reads as orchestration. Wrapped both the class name and the method filter in `Pattern.quote()` to kill the regex-metacharacter edge case, and dropped the now-redundant `.*` wrappers since JMH matches with `.find()`. Pushed in `6ce02f65`.

**Review by LihaoLiuXs — APPROVED** [review]
LGTM. Not related to this PR: can we add a test case with large files for the directory operations? Currently we only test with a 50 MB file, but some clients have files larger than 1 GB.

**Comment by iamabhilaksh** [comment]
> LGTM. Not related to this PR: can we add a test case with large files for the directory operations? Currently we only test with a 50 MB file, but some clients have files larger than 1 GB.

Yes definitely @LihaoLiuXs. 1 GB file size in benchmark tests makes it extremely slow, susceptible to network timeout and memory issues. The idea is to run every variant of file size : XS, S, M, L, XL on the pods itself when the benchmark pipeline is completely established.

**Review by sharatchandrag — APPROVED** [review]
